# Acquire

**Goal: Your goal is to predict the values of single unit properties using the obervations from 2017.**

**import**

In [29]:
# data manipulation
import pandas as pd
import numpy as np

# data visualization
import matplotlib.pyplot as plt
import seaborn as sns

# data separation/transformation
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

# system manipulation
import sys
sys.path.append("./util_")
import acquire_
import prepare_

# other
import warnings
warnings.filterwarnings("ignore")


**get data**

In [30]:
# sql query
query = """
SELECT bedroomcnt, 
        bathroomcnt,
        calculatedfinishedsquarefeet,
        taxvaluedollarcnt,
        yearbuilt,
        taxamount,
        fips     
FROM properties_2017
WHERE propertylandusetypeid = 261; -- Single family home
"""

In [31]:
# get data from codeup database
zillow, q = acquire_.get_codeup_sql_data_(db_name="zillow", query=query,fileName="zillow_single_family")


**Understand data**

In [32]:
zillow.head()

bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  taxvaluedollarcnt  \
0         0.0          0.0                           NaN            27516.0   
1         0.0          0.0                           NaN               10.0   
2         0.0          0.0                           NaN               10.0   
3         0.0          0.0                           NaN             2108.0   
4         4.0          2.0                        3633.0           296425.0   

   yearbuilt  taxamount    fips  
0        NaN        NaN  6037.0  
1        NaN        NaN  6037.0  
2        NaN        NaN  6037.0  
3        NaN     174.21  6037.0  
4     2005.0    6941.39  6037.0

In [33]:
zillow.shape

(2152863, 7)

In [34]:
zillow.columns

Index(['bedroomcnt', 'bathroomcnt', 'calculatedfinishedsquarefeet',
       'taxvaluedollarcnt', 'yearbuilt', 'taxamount', 'fips'],
      dtype='object')

In [35]:
zillow.info(show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2152863 entries, 0 to 2152862
Data columns (total 7 columns):
 #   Column                        Non-Null Count    Dtype  
---  ------                        --------------    -----  
 0   bedroomcnt                    2152852 non-null  float64
 1   bathroomcnt                   2152852 non-null  float64
 2   calculatedfinishedsquarefeet  2144379 non-null  float64
 3   taxvaluedollarcnt             2152370 non-null  float64
 4   yearbuilt                     2143526 non-null  float64
 5   taxamount                     2148421 non-null  float64
 6   fips                          2152863 non-null  float64
dtypes: float64(7)
memory usage: 115.0 MB


In [36]:
# count all the null values
zillow.isna().sum().values.sum() / len(zillow)

0.01058032954256727

In [37]:
zillow.describe()

bedroomcnt   bathroomcnt  calculatedfinishedsquarefeet  \
count  2.152852e+06  2.152852e+06                  2.144379e+06   
mean   3.287196e+00  2.230688e+00                  1.862855e+03   
std    9.547544e-01  9.992796e-01                  1.222125e+03   
min    0.000000e+00  0.000000e+00                  1.000000e+00   
25%    3.000000e+00  2.000000e+00                  1.257000e+03   
50%    3.000000e+00  2.000000e+00                  1.623000e+03   
75%    4.000000e+00  3.000000e+00                  2.208000e+03   
max    2.500000e+01  3.200000e+01                  9.525760e+05   

       taxvaluedollarcnt     yearbuilt     taxamount          fips  
count       2.152370e+06  2.143526e+06  2.148421e+06  2.152863e+06  
mean        4.618962e+05  1.960950e+03  5.634866e+03  6.048377e+03  
std         6.996760e+05  2.216220e+01  8.178910e+03  2.043329e+01  
min         1.000000e+00  1.801000e+03  1.850000e+00  6.037000e+03  
25%         1.881702e+05  1.949000e+03  2.534980e+03  6.037000e+03  
50%         3.276710e+05  1.958000e+03  4.108950e+03  6.037000e+03  
75%         5.345270e+05  1.976000e+03  6.414320e+03  6.059000e+03  
max         9.842891e+07  2.016000e+03  1.337756e+06  6.111000e+03

In [38]:
# count of numeric columns and object columns
numerics = len(zillow.select_dtypes("number").columns)
objects = len(zillow.select_dtypes("object").columns)

print("Numeric col count:", numerics)
print("object col count:", objects)

Numeric col count: 7
object col count: 0


**What I see:**

- I have 2152863 rows and 7 columns
- 7 of the 29 columns are numric while 0 of them are string object colums
- 22778 total null count (1% of the data)
- I also see the descriptive statistics of my data

# Prepare

**Rename columns**

In [39]:
# rename dataframe columns
zillow = zillow.rename(columns={"bedroomcnt":"bedrooms",
                       "bathroomcnt":"bathrooms",
                       "calculatedfinishedsquarefeet":"sqr_feet",
                      "taxvaluedollarcnt":"tax_value",
                      "yearbuilt":"year_built",
                      "taxamount":"tax_amount",
                      "fips":"county"})

**drop nulls**

In [40]:
# drop all nulls in the dataframe
zillow = zillow.dropna()

**convert data type**

In [41]:
# convert data type from float to int
zillow.bedrooms = zillow.bedrooms.astype(int)
zillow.year_built = zillow.year_built.astype(int)

**Remove duplicated rows**

In [42]:
# remove the duplocated rows
zillow = zillow.drop_duplicates(keep="first")

**Remove outliers**

This is done using box plot in the explore phase. only looking at the training data.

In [43]:
# remove outliers
zillow = zillow[zillow.bedrooms < 5]
zillow = zillow[zillow.bathrooms < 5]
zillow = zillow[(zillow.year_built >= 1910) & (zillow.year_built <= 2017)]
zillow = zillow[(zillow.sqr_feet > 500) & (zillow.sqr_feet <= 3800)]
zillow = zillow[zillow.tax_amount <= 20000]

**feature transformation**

In [44]:
# Rename the unique values in fips to county names
zillow.county = zillow.county.astype(str).str.replace("6037.0","Los Angeles").str.replace("6059.0","Orange").str.replace("6111.0","Ventura")


**Create dummies**

In [45]:
# create dummie variables
dummies = pd.get_dummies(zillow.county)

# clean dummie column names
dummies_col = dummies.columns.str.replace(" ", "_").str.lower()

dummies.head(2)

Los Angeles  Orange  Ventura
4            1       0        0
6            1       0        0

In [46]:
# make a copy of my original data frame
zillow_encoded_scaled = zillow.copy()

# add dummies to my data frame
zillow_encoded_scaled[dummies_col] = dummies
zillow_encoded_scaled.head(2)

bedrooms  bathrooms  sqr_feet  tax_value  year_built  tax_amount  \
4         4        2.0    3633.0   296425.0        2005     6941.39   
6         3        4.0    1620.0   847770.0        2011    10244.94   

        county  los_angeles  orange  ventura  
4  Los Angeles            1       0        0  
6  Los Angeles            1       0        0

## Split

In [47]:
# split the data into training, validation and testing sets
train, validate, test = prepare_.split_data_(df=zillow_encoded_scaled,
                    test_size=0.2, 
                     validate_size=0.2,
                    stratify_col= "county",
                     random_state=95)
(train.shape, validate.shape, test.shape)

((1120943, 10), (373648, 10), (373648, 10))

**Scale data**

In [48]:
# scalable features
features_to_scale = train[['bedrooms','bathrooms','sqr_feet','year_built','tax_amount']]

features_to_scale.head(2)

bedrooms  bathrooms  sqr_feet  year_built  tax_amount
615369          3        2.0    1532.0        1954     4037.13
1437240         2        2.0    1478.0        1929     5021.84

In [49]:
# build a scaling object
scaler = MinMaxScaler()

# Note that we only call .fit with the training data,
# but we use .transform to apply the scaling to all the data splits.
# y Target values: this is not being changed in any way(no predictions are being made)
x_train_scaled = scaler.fit_transform(X=features_to_scale,)

# transfrom the validate and test using the minMax object
x_val_scaled = scaler.transform(X=validate[features_to_scale.columns])
x_test_scaled = scaler.transform(X=test[features_to_scale.columns])


In [50]:
x_train_scaled

array([[0.75      , 0.44444444, 0.31251895, 0.41509434, 0.20111035],
       [0.5       , 0.44444444, 0.29615035, 0.17924528, 0.25039237],
       [0.75      , 0.44444444, 0.44710518, 0.5       , 0.11126754],
       ...,
       [0.5       , 0.44444444, 0.15792664, 0.31132075, 0.19675624],
       [0.75      , 0.66666667, 0.42891785, 0.80188679, 0.2971425 ],
       [1.        , 0.88888889, 0.79084571, 0.91509434, 0.51615375]])

**Convert scaled features to dataframe**

In [51]:
# New _ariable mames to add to data
new_scale_col = []
for i in features_to_scale.columns:
    new_scale_col.append(f"{i}_scaled")

new_scale_col

['bedrooms_scaled',
 'bathrooms_scaled',
 'sqr_feet_scaled',
 'year_built_scaled',
 'tax_amount_scaled']

In [52]:
# convert to dataframe
x_train_scaled = pd.DataFrame(x_train_scaled)
x_val_scaled = pd.DataFrame(x_val_scaled)
x_test_scaled = pd.DataFrame(x_test_scaled)

# add new column names back to the data frame
x_train_scaled[new_scale_col] = x_train_scaled
x_val_scaled[new_scale_col] = x_val_scaled
x_test_scaled[new_scale_col] = x_test_scaled

# remove redundent columns from new dataframe
x_train_scaled = x_train_scaled[new_scale_col]
x_val_scaled = x_val_scaled[new_scale_col]
x_test_scaled = x_test_scaled[new_scale_col]

x_train_scaled.head()

bedrooms_scaled  bathrooms_scaled  sqr_feet_scaled  year_built_scaled  \
0             0.75          0.444444         0.312519           0.415094   
1             0.50          0.444444         0.296150           0.179245   
2             0.75          0.444444         0.447105           0.500000   
3             0.75          0.444444         0.339194           0.745283   
4             0.75          0.222222         0.194908           0.443396   

   tax_amount_scaled  
0           0.201110  
1           0.250392  
2           0.111268  
3           0.174100  
4           0.146260

In [53]:
x_train_scaled.shape

(1120943, 5)

**Add scaled columns to a copy of original data**

In [54]:
# train[x_train_scaled.columns] = x_train_scaled
train = train.reset_index(drop=True)
validate = validate.reset_index(drop=True)
test = test.reset_index(drop=True)

# Concatenate the DataFrames horizontally
train = pd.concat([train, x_train_scaled], axis=1, verify_integrity=True)
validate = pd.concat([validate, x_val_scaled], axis=1, verify_integrity=True)
test = pd.concat([test, x_test_scaled], axis=1, verify_integrity=True)

# drop the original columns
train = train.drop(columns=features_to_scale)
validate = validate.drop(columns=features_to_scale)
test = test.drop(columns=features_to_scale)

In [55]:
train.head()

tax_value       county  los_angeles  orange  ventura  bedrooms_scaled  \
0   326232.0  Los Angeles            1       0        0             0.75   
1   336547.0  Los Angeles            1       0        0             0.50   
2   138770.0  Los Angeles            1       0        0             0.75   
3   275868.0  Los Angeles            1       0        0             0.75   
4   219404.0  Los Angeles            1       0        0             0.75   

   bathrooms_scaled  sqr_feet_scaled  year_built_scaled  tax_amount_scaled  
0          0.444444         0.312519           0.415094           0.201110  
1          0.444444         0.296150           0.179245           0.250392  
2          0.444444         0.447105           0.500000           0.111268  
3          0.444444         0.339194           0.745283           0.174100  
4          0.222222         0.194908           0.443396           0.146260

**Save split**

In [56]:
# save created data frames into csv
prepare_.save_split_data_(original_df=zillow,
                          encoded_scaled_df=zillow_encoded_scaled, 
                          train=train, 
                          validate=validate, 
                          test=test,
                         test_size=0.2,
                         stratify_col= "county",
                         random_state=95)

'SIX data sets saved as .csv'

### Actions Taken

- Remove all nulls (1% of the data)
- Remove duplicated rows.
- convert data type from float to int (bedrooms, bathrooms)
- remove outliers
- replace the fips code with county names and Encode county column.
- Split data into train, validate, and test. (`60/20/20 ` split)
- scale the humeric categorical and continuous variable and extract a copy of the original data frame.
    - `bedrooms, bathrooms, sqr_feet, year_built, tax_amount`